ir/js/src/main/scala/org/scalajs/ir/SHA1.scala

```diff
@@ -0,0 +1,398 @@
+      @inline def R4(v: IntRef, w: IntRef, x: IntRef, y: IntRef, z: IntRef, i: Int): Unit = {
+        z() += (w() ^ x() ^ y()) + blk(i) + 0xca62c1d6 + rol(v(), 5)
+        w() = rol(w(), 30)
+      }
+
+      // Copy state to working vars
+      val a = new IntRef(state0)
+      val b = new IntRef(state1)
+      val c = new IntRef(state2)
+      val d = new IntRef(state3)
+      val e = new IntRef(state4)
+
+      // 4 rounds of 20 operations each. Loop unrolled. (verbatim copy from C code)
+      R0(a, b, c, d, e, 0)
+      R0(e, a, b, c, d, 1)
+      R0(d, e, a, b, c, 2)
+      R0(c, d, e, a, b, 3)
+      R0(b, c, d, e, a, 4)
+      R0(a, b, c, d, e, 5)
+      R0(e, a, b, c, d, 6)
+      R0(d, e, a, b, c, 7)
+      R0(c, d, e, a, b, 8)
+      R0(b, c, d, e, a, 9)
+      R0(a, b, c, d, e, 10)
+      R0(e, a, b, c, d, 11)
+      R0(d, e, a, b, c, 12)
+      R0(c, d, e, a, b, 13)
+      R0(b, c, d, e, a, 14)
+      R0(a, b, c, d, e, 15)
+      R1(e, a, b, c, d, 16)
+      R1(d, e, a, b, c, 17)
+      R1(c, d, e, a, b, 18)
+      R1(b, c, d, e, a, 19)
+      R2(a, b, c, d, e, 20)
+      R2(e, a, b, c, d, 21)
+      R2(d, e, a, b, c, 22)
+      R2(c, d, e, a, b, 23)
+      R2(b, c, d, e, a, 24)
+      R2(a, b, c, d, e, 25)
+      R2(e, a, b, c, d, 26)
+      R2(d, e, a, b, c, 27)
+      R2(c, d, e, a, b, 28)
+      R2(b, c, d, e, a, 29)
+      R2(a, b, c, d, e, 30)
+      R2(e, a, b, c, d, 31)
+      R2(d, e, a, b, c, 32)
+      R2(c, d, e, a, b, 33)
+      R2(b, c, d, e, a, 34)
+      R2(a, b, c, d, e, 35)
+      R2(e, a, b, c, d, 36)
+      R2(d, e, a, b, c, 37)
+      R2(c, d, e, a, b, 38)
+      R2(b, c, d, e, a, 39)
+      R3(a, b, c, d, e, 40)
+      R3(e, a, b, c, d, 41)
+      R3(d, e, a, b, c, 42)
+      R3(c, d, e, a, b, 43)
+      R3(b, c, d, e, a, 44)
+      R3(a, b, c, d, e, 45)
+      R3(e, a, b, c, d, 46)
+      R3(d, e, a, b, c, 47)
+      R3(c, d, e, a, b, 48)
+      R3(b, c, d, e, a, 49)
+      R3(a, b, c, d, e, 50)
+      R3(e, a, b, c, d, 51)
+      R3(d, e, a, b, c, 52)
+      R3(c, d, e, a, b, 53)
+      R3(b, c, d, e, a, 54)
+      R3(a, b, c, d, e, 55)
+      R3(e, a, b, c, d, 56)
+      R3(d, e, a, b, c, 57)
+      R3(c, d, e, a, b, 58)
+      R3(b, c, d, e, a, 59)
+      R4(a, b, c, d, e, 60)
+      R4(e, a, b, c, d, 61)
+      R4(d, e, a, b, c, 62)
+      R4(c, d, e, a, b, 63)
+      R4(b, c, d, e, a, 64)
+      R4(a, b, c, d, e, 65)
+      R4(e, a, b, c, d, 66)
+      R4(d, e, a, b, c, 67)
+      R4(c, d, e, a, b, 68)
+      R4(b, c, d, e, a, 69)
+      R4(a, b, c, d, e, 70)
+      R4(e, a, b, c, d, 71)
+      R4(d, e, a, b, c, 72)
+      R4(c, d, e, a, b, 73)
+      R4(b, c, d, e, a, 74)
+      R4(a, b, c, d, e, 75)
+      R4(e, a, b, c, d, 76)
+      R4(d, e, a, b, c, 77)
+      R4(c, d, e, a, b, 78)
+      R4(b, c, d, e, a, 79)
+
+      // Add the working vars back into `state`
+      state0 += a()
+      state1 += b()
+      state2 += c()
+      state3 += d()
+      state4 += e()
```

Comment:
I do not understand why `+=` is correct here: Aren't the boxes (`a` - `z`) essentially stack allocated copies? So the put the state back, why do we add and not assign?

Reply:
It's just how SHA-1 works. We initialize `a b c d e` from the state, apply many rounds of transformations to those 5 variables, and then *add* their values back to the state. See for example https://en.wikipedia.org/wiki/SHA-1#SHA-1_pseudocode

a b c d e are distinct from the state in the original algorithm. They're not just stack-allocated copies of the state made for Scala purposes.